dpcpp/base/config.hpp

```diff
@@ -72,6 +76,68 @@ struct config {
 };
 
 
+/**
+ * KCFG_1D provides the usual way to embed information from workgroup size and
+ * sub_group size. We consider the workgroup size up to 4096 which requires 13
+ * bits to store it and sub_group size up to 64 which requires 7 bits to store
+ * it.
+ */
+using KCFG_1D = ConfigSet<11, 7>;
+
+
+template <uint32 block, uint32 subgroup>
+struct device_config {
+    static constexpr uint32 block_size = block;
+    static constexpr uint32 subgroup_size = subgroup;
+    static constexpr uint32 encode = KCFG_1D::encode(block_size, subgroup_size);
+};
+
+/**
+ * encode_list base type
+ *
+ * @tparam T  the input template
+ */
+template <typename T>
+struct encode_list {};
+
+/**
+ * encode_list specializes for the type_list. It will convert the each type to
+ * encoded information.
+ *
+ * @tparam T  the input template
+ */
+template <typename... Types>
+struct encode_list<syn::type_list<Types...>> {
+    using type = syn::value_list<uint32, Types::encode...>;
+};
```

Comment:
This should work, so you don't need the specialization:
```suggestion
template <typename... Types>
struct encode_list {
    using type = syn::value_list<uint32, Types::encode...>;
};
```

Reply:
ah, you are right. I somehow think I need to use `type_list` to wrap the type list always.

Reply:
I need to use `type_list<type...>` as input.
I create a `type_list` type `using dcfg_type_list_t =  type_list<device_config<.>...>;` and then use it to create its value_list varient `using dcfg_list_t = encode<dcfg_type_list>::type;`

Reply:
OK, makes sense.